Adopt standards renames for declarative web push, and enable it

beidson opened the pull request:
#### 79bda7ce145cbd5c2e4b786e82920e9e587a37cd
<pre>
Adopt standards renames for declarative web push, and enable it
<a href="https://rdar.apple.com/141082392">rdar://141082392</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=284214">https://bugs.webkit.org/show_bug.cgi?id=284214</a>

Reviewed by Per Arne Vollan.

Recent edits to the standards PRs have renamed a property, and moved some properties from
a new type of push event object onto the existing PushEvent.

Adopt those changes and enable.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/notifications/Notification.cpp:
(WebCore::Notification::create):
(WebCore::Notification::Notification):
(WebCore::Notification::data const):
* Source/WebCore/Modules/notifications/Notification.h:
* Source/WebCore/Modules/notifications/Notification.idl:
* Source/WebCore/Modules/notifications/NotificationData.cpp:
(WebCore::NotificationData::isolatedCopy const):
(WebCore::NotificationData::isolatedCopy):
* Source/WebCore/Modules/notifications/NotificationData.h:
* Source/WebCore/Modules/notifications/NotificationDataCocoa.mm:
(WebCore::NotificationData::dictionaryRepresentation const):
* Source/WebCore/Modules/notifications/NotificationOptions.idl:
* Source/WebCore/Modules/notifications/NotificationPayload.cpp:
(WebCore::NotificationPayload::fromNotificationData):
(WebCore::NotificationPayload::toNotificationData const):
* Source/WebCore/Modules/push-api/PushEvent.cpp:
(WebCore::PushEvent::create):
(WebCore::PushEvent::PushEvent):
(WebCore::PushEvent::notification):
(WebCore::PushEvent::appBadge):
(WebCore::PushEvent::updatedNotificationData const):
* Source/WebCore/Modules/push-api/PushEvent.h:
* Source/WebCore/Modules/push-api/PushEvent.idl:
* Source/WebCore/Modules/push-api/PushNotificationEvent.cpp: Removed.
* Source/WebCore/Modules/push-api/PushNotificationEvent.h: Removed.
* Source/WebCore/Modules/push-api/PushNotificationEvent.idl: Removed.
* Source/WebCore/Modules/push-api/PushNotificationEventInit.h: Removed.
* Source/WebCore/Modules/push-api/PushNotificationEventInit.idl: Removed.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventInterfaces.in:
* Source/WebCore/page/WorkerNavigator.cpp:
(WebCore::WorkerNavigator::setAppBadge):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
(WebCore::ServiceWorkerGlobalScope::dispatchPushEvent):
(WebCore::ServiceWorkerGlobalScope::dispatchDeclarativePushEvent):
(WebCore::ServiceWorkerGlobalScope::clearDeclarativePushEvent):
(WebCore::ServiceWorkerGlobalScope::dispatchPushNotificationEvent): Deleted.
(WebCore::ServiceWorkerGlobalScope::clearPushNotificationEvent): Deleted.
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.h:
* Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
(WebCore::ServiceWorkerRegistration::showNotification):
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::queueTaskToFirePushEvent):
(WebCore::ServiceWorkerThread::queueTaskToFireDeclarativePushEvent):
(WebCore::ServiceWorkerThread::queueTaskToFirePushNotificationEvent): Deleted.
* Source/WebCore/workers/service/context/ServiceWorkerThread.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _processWebCorePersistentNotificationClick:completionHandler:]):

Canonical link: <a href="https://commits.webkit.org/287525@main">https://commits.webkit.org/287525@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c21bc4dcb3b27adb82a27a8ea9758f20b0dab9ae

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79985 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58982 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33383 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84499 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30959 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82093 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68046 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7278 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62520 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20346 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83052 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52583 "Found 1 new test failure: media/audioSession/getUserMedia.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72844 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42829 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49920 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26997 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29421 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/72959 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71040 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27488 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85931 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/79044 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7201 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5067 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70792 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7378 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68685 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70036 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17446 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14031 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12966 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/101450 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7165 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12699 "Found 1 new failure in JSPushEvent.cpp and found 1 fixed file: JSWebCodecsEncodedVideoChunk.cpp") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24737 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-no-cjit (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7012 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10524 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8816 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebCore/Modules/push-api/PushEvent.h:
```diff
@@ -37,14 +39,32 @@ class PushEvent final : public ExtendableEvent {
+    std::optional<uint64_t> m_proposedAppBadge;
+
+    RefPtr<Notification> m_updatedNotification;
+    std::optional<std::optional<uint64_t>> m_updatedAppBadge;
```
I am curious why this needs to be of the type `std::optional<std::optional<uint64_t>>`?

beidson replied:
LOL - Second time to have talked about this, first when originally implemented:

An app badge is "an optional uint64_t" - You either set it to a number, or you set it to nothing to clear it.
And in this code, we need to capture "has the app badge been overridden at all?"

I could create a `struct AppBadge` which has a `uint64_t` member and a `bool m_isNull` member, and then we'd have `std::optional<AppBadge>` here instead, and the code would read nicer

beidson replied:
In fact if I have to update this patch again for EWS, I might add `typedef std::optional<uint64_t> AppBadge` for readability.

pvollan reviewed (approved):
R=me.

beidson committed:
Adopt standards renames for declarative web push, and enable it
rdar://141082392
https://bugs.webkit.org/show_bug.cgi?id=284214

Reviewed by Per Arne Vollan.

Recent edits to the standards PRs have renamed a property, and moved some properties from
a new type of push event object onto the existing PushEvent.

Adopt those changes and enable.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/notifications/Notification.cpp:
(WebCore::Notification::create):
(WebCore::Notification::Notification):
(WebCore::Notification::data const):
* Source/WebCore/Modules/notifications/Notification.h:
* Source/WebCore/Modules/notifications/Notification.idl:
* Source/WebCore/Modules/notifications/NotificationData.cpp:
(WebCore::NotificationData::isolatedCopy const):
(WebCore::NotificationData::isolatedCopy):
* Source/WebCore/Modules/notifications/NotificationData.h:
* Source/WebCore/Modules/notifications/NotificationDataCocoa.mm:
(WebCore::NotificationData::dictionaryRepresentation const):
* Source/WebCore/Modules/notifications/NotificationOptions.idl:
* Source/WebCore/Modules/notifications/NotificationPayload.cpp:
(WebCore::NotificationPayload::fromNotificationData):
(WebCore::NotificationPayload::toNotificationData const):
* Source/WebCore/Modules/push-api/PushEvent.cpp:
(WebCore::PushEvent::create):
(WebCore::PushEvent::PushEvent):
(WebCore::PushEvent::notification):
(WebCore::PushEvent::appBadge):
(WebCore::PushEvent::updatedNotificationData const):
* Source/WebCore/Modules/push-api/PushEvent.h:
* Source/WebCore/Modules/push-api/PushEvent.idl:
* Source/WebCore/Modules/push-api/PushNotificationEvent.cpp: Removed.
* Source/WebCore/Modules/push-api/PushNotificationEvent.h: Removed.
* Source/WebCore/Modules/push-api/PushNotificationEvent.idl: Removed.
* Source/WebCore/Modules/push-api/PushNotificationEventInit.h: Removed.
* Source/WebCore/Modules/push-api/PushNotificationEventInit.idl: Removed.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventInterfaces.in:
* Source/WebCore/page/WorkerNavigator.cpp:
(WebCore::WorkerNavigator::setAppBadge):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
(WebCore::ServiceWorkerGlobalScope::dispatchPushEvent):
(WebCore::ServiceWorkerGlobalScope::dispatchDeclarativePushEvent):
(WebCore::ServiceWorkerGlobalScope::clearDeclarativePushEvent):
(WebCore::ServiceWorkerGlobalScope::dispatchPushNotificationEvent): Deleted.
(WebCore::ServiceWorkerGlobalScope::clearPushNotificationEvent): Deleted.
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.h:
* Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
(WebCore::ServiceWorkerRegistration::showNotification):
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::queueTaskToFirePushEvent):
(WebCore::ServiceWorkerThread::queueTaskToFireDeclarativePushEvent):
(WebCore::ServiceWorkerThread::queueTaskToFirePushNotificationEvent): Deleted.
* Source/WebCore/workers/service/context/ServiceWorkerThread.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _processWebCorePersistentNotificationClick:completionHandler:]):

Canonical link: https://commits.webkit.org/287525@main

rkirsling commented on Source/WebCore/Modules/push-api/PushEvent.h:
```diff
@@ -26,6 +26,8 @@
 #pragma once
 
 #include "ExtendableEvent.h"
+#include "Notification.h"
```
Notification.h is wrapped in ENABLE(NOTIFICATIONS), but this file has no feature guards whatsoever. How do you mean for the added code to be guarded?
(PlayStation is currently `!ENABLE(NOTIFICATIONS)` and cannot build.)

darinadler replied:
I looked at this with @rkirsling and my take is that a lot this new `PushEvent` stuff needs to be wrapped in `ENABLE(DECLARATIVE_WEB_PUSH)`; it won’t be terribly difficult, but it’s not trivial. Should re-land with that done

beidson replied:
That was missed since we have no such EWS coverage.
The additions to PushEvent.h/cpp/idl that are conditioned on that need to be guarded.
Without the failing build in front of me I'm not sure I could make meaningful progress on that - Is there a link to a build outside our EWS infrastructure that I could see?

rkirsling replied:
You'd be able to look at the PlayStation post-merge builders, though it'd probably be easier to temporarily set ENABLE_NOTIFICATIONS to 0 in PlatformEnable.h. If it had just been a matter of missing a spot or two, I'd already have a fix landed, but the edit to the existing `PushEvent` constructor is where things get less trivial.

rkirsling replied:
Here's one possible fix:
https://gist.github.com/rkirsling/0f4c2ed4aa514142f75f7274d70b0744

beidson replied:
> easier to temporarily set ENABLE_NOTIFICATIONS to 0 in PlatformEnable.h...

This may not be as trivial as you suggest for a platform that has all of the related features enabled as well!
But I'll try.

rkirsling replied:
True, it appears that the actual way to do this on Mac is to add
```cpp
#define ENABLE_DECLARATIVE_WEB_PUSH 0
#define ENABLE_NOTIFICATION_EVENT 0
#define ENABLE_NOTIFICATIONS 0
#define ENABLE_WEB_PUSH_NOTIFICATIONS 0
```
at the top of PlatformEnableCocoa.h (or before its `#include` in PlatformEnable.h).

It would also be possible to do a best guess without building anything, which might lead you to make the exact fix patch above. A best guess would have been enough to avoid a revert here.

beidson replied:
IMHO, that fix isn't right.

Because the problem isn't that Playstation doesn't have DECLARATIVE_WEB_PUSH off...  It's that it has NOTIFICATIONS off.

And on the topic of best guesses - There's a reason we have EWS, and we're used to bouncing best guesses off of it to qualify (e.g. we're constantly keeping linux builds working with the help of EWS)

beidson replied:
Patch focused on ENABLE_NOTIFICATIONS on its way.

beidson replied:
The *REAL* real fix here is much more extension enforcement of ENABLE_WEB_PUSH_NOTIFICATIONS, because no port should have the PushEvent at all if it doesn't have basic web push on.
But that patch quickly grew much larger in scope compared to the actual work being done here.

I have that patch stashed, and if/when we get a Playstation EWS I'll follow up with it.

rkirsling replied:
I had just replaced `ENABLE(NOTIFICATIONS)` with `ENABLE(DECLARATIVE_WEB_PUSH)` based on @darinadler's suggestion; I'm certainly fine with the former.

> And on the topic of best guesses - There's a reason we have EWS

No, my entire point here is that I would've happily fixed our build _without_ reverting you if it boiled down to a simple oversight. Changing the constructor of a class which is built unconditionally on all platforms to add a _non-final_ parameter of a type that exists conditionally on some platforms made me feel like your patch was based on expectations that do not hold, and therefore I needed you to be involved in the fix.
